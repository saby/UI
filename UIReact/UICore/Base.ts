export { default as Control}  from './_base/Control';

export {IControlOptions, TemplateFunction} from './_base/interfaces';
