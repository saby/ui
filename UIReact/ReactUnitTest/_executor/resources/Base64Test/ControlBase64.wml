<div class="ReactUnitTest-Executor__base64">
  <div name="onlyBase64" attr:style="background-image: url({{ _value }})"></div>
  <div name="otherPropBeforeBase64" attr:style="color: red; background-image: url({{ _value }})"></div>
  <div name="otherAfterBeforeBase64" attr:style="background-image: url({{ _value }}); color: red;"></div>
  <div name="otherPropsAroundBase64" attr:style="color: red; background-image: url({{ _value }}); margin: 8px;"></div>
</div>
