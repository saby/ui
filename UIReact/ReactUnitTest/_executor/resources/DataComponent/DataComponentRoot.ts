import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_executor/resources/DataComponent/DataComponentRoot';

export default class DataComponentRoot extends Control {
    _template: TemplateFunction = template;
}
