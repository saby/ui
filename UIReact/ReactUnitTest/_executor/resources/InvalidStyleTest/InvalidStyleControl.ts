import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_executor/resources/InvalidStyleTest/InvalidStyleControl';

export default class InvalidStyleControl extends Control {
    _template: TemplateFunction = template;
    protected maxWidth: number = 120;
    resetMaxWidth(): void {
        this.maxWidth = undefined;
    }
}
