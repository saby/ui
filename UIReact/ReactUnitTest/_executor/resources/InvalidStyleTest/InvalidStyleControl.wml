<div>
   <div id="divWithStyle" style="max-width: {{ maxWidth }}px;"></div>
</div>
