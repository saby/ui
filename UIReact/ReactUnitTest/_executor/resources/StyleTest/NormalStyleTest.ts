import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_executor/resources/StyleTest/NormalStyleTest';

export default class NormalStyleTest extends Control {
    _template: TemplateFunction = template;
    protected gridColumns: number = 1;
}
