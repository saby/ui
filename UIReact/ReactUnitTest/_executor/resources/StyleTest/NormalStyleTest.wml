<div>
    <div style="grid-template-columns: {{ gridColumns }}fr;"></div>
</div>
