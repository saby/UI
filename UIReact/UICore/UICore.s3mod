<?xml version="1.0" encoding="UTF-8" ?>
<ui_module id="548e4ee9-10cd-4227-9e3c-0811958f979c" name="UICore" package="Платформа" responsible="Шипин А.А." required="1" version="1.00">

  <description>Библиотека для работы ядра через React. Имеет общее API с UIInferno/UICore</description>

  <depends>
    <module id="0f15ac6b-86b2-4265-9da1-05eac93ebdf6" name="UICommon" />
  </depends>

  <load_after/>

  <parameters/>

  <resources/>

  <preload/>

</ui_module>
