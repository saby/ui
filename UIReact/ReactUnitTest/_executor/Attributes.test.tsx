/**
 * @jest-environment jsdom
 */
import { unmountComponentAtNode } from 'react-dom';

import { createSandbox, SinonSandbox } from 'sinon';
import { act } from 'react-dom/test-utils';
import { delay } from 'Types/function';

import { Control } from 'UICore/Base';

import ControlBase64, { styles } from './resources/Base64Test/ControlBase64';

import ControlSVG from './resources/SvgTest/ControlSVG';
import InvalidStyleControl from './resources/InvalidStyleTest/InvalidStyleControl';
import IeStyleTest from './resources/StyleTest/IeStyleTest';
import NormalStyleTest from './resources/StyleTest/NormalStyleTest';
import DataComponentRoot from './resources/DataComponent/DataComponentRoot';
import WmlAttrsWasabyRoot from './resources/WmlAttrs/WasabyRoot';

const creator = Control.createControl;

describe('Markup attributes', () => {
    let errors = [];
    const mockedError = (output) => {
        return errors.push(output);
    };
    let container;
    let sandbox: SinonSandbox;
    let clock;

    beforeEach(() => {
        sandbox = createSandbox();
        clock = sandbox.useFakeTimers();
        sandbox.stub(console, 'error').callsFake(mockedError);
        sandbox.stub(window, 'requestAnimationFrame').callsFake(setTimeout);
        container = document.createElement('div');
        document.body.appendChild(container);
    });

    afterEach(() => {
        errors = [];
        clock.restore();
        sandbox.restore();
        unmountComponentAtNode(container);
        container.remove();
        container = null;
    });

    function tick(duration: number): void {
        act(() => {
            clock.tick(duration);
        });
    }

    async function tickWasabyUpdate(): Promise<void> {
        return act(async () => {
            await Promise.resolve();

            // За время requestAnimationFrame отвечает jsdom.
            // Можно было бы с запасом сделать clock.tick(100), но так точнее.
            await new Promise<void>((resolve) => {
                let resolved: boolean = false;
                // forceUpdate занимает 2 delay
                delay(() => {
                    delay(() => {
                        resolved = true;
                        resolve();
                    });
                });
                while (!resolved) {
                    clock.tick(1);
                }
            });
        });
    }

    it('Построение svg с xlink:href', async () => {
        // region Setup
        let instance;
        act(() => {
            instance = creator(ControlSVG, {}, container);
        });
        tick(0);
        // endregion

        const numberOfMatches = errors.filter((x) => {
            return x.indexOf('Invalid DOM property') > -1;
        }).length;
        expect(numberOfMatches).toBe(0);
    });

    it('Валидация строки style', async () => {
        // region Setup

        // Chrome и IE не присваивает невалидные значения CSSStyleDeclaration, jsdom - присваивает.
        // Мокнем конкретный сценарий.
        sandbox
            .stub(CSSStyleDeclaration.prototype, 'maxWidth')
            .get(function getMaxWidth(): string {
                return this.maxWidthValue || '';
            })
            .set(function setMaxWidth(a: string): void {
                if (a === 'px') {
                    return;
                }
                this.maxWidthValue = a;
            });
        let instance: InvalidStyleControl;
        act(() => {
            instance = creator(InvalidStyleControl, {}, container);
        });
        await tickWasabyUpdate();

        // endregion

        instance.resetMaxWidth();
        tick(2);
        expect(container).toMatchSnapshot();
    });

    it('Валидация строки style ie', async () => {
        // region Setup

        let instance: IeStyleTest;
        act(() => {
            instance = creator(IeStyleTest, {}, container);
        });
        await tickWasabyUpdate();

        // endregion
        tick(2);
        expect(container).toMatchSnapshot();
    });

    it('Валидация строки style normal', async () => {
        // region Setup
        let instance: NormalStyleTest;
        act(() => {
            instance = creator(NormalStyleTest, {}, container);
        });
        await tickWasabyUpdate();

        // endregion
        tick(2);
        expect(container).toMatchSnapshot();
    });

    it('Строка style c base64 внутри', () => {
        // region Setup
        let instance: ControlBase64;
        act(() => {
            instance = creator(ControlBase64, {}, container) as ControlBase64;
        });
        tick(0);
        // endregion

        /** На конечном DOM элементе стили должны быть с base64 */
        expect(instance._children.onlyBase64.style.backgroundImage).toBe(
            styles.backgroundImage
        );

        expect(instance._children.otherPropBeforeBase64.style.color).toBe(
            styles.color
        );
        expect(
            instance._children.otherPropBeforeBase64.style.backgroundImage
        ).toBe(styles.backgroundImage);

        expect(
            instance._children.otherAfterBeforeBase64.style.backgroundImage
        ).toBe(styles.backgroundImage);
        expect(instance._children.otherAfterBeforeBase64.style.color).toBe(
            styles.color
        );

        expect(instance._children.otherPropsAroundBase64.style.color).toBe(
            styles.color
        );
        expect(
            instance._children.otherPropsAroundBase64.style.backgroundImage
        ).toBe(styles.backgroundImage);
        expect(instance._children.otherPropsAroundBase64.style.margin).toBe(
            styles.margin
        );
    });

    it('data-component ставиться на корректный div', async () => {
        // region Setup
        let instance;
        act(() => {
            instance = creator(DataComponentRoot, {}, container);
        });
        tick(0);
        // endregion
        await tickWasabyUpdate();

        expect(container).toMatchSnapshot();
    });
});

describe('attr: в wml', () => {
    let container: HTMLDivElement;

    beforeEach(() => {
        container = document.createElement('div');
        document.body.appendChild(container);

        jest.useFakeTimers();
        jest.spyOn(window, 'requestAnimationFrame').mockImplementation(
            setTimeout
        );
    });

    afterEach(() => {
        jest.useRealTimers();
        unmountComponentAtNode(container);
        container.remove();
        jest.restoreAllMocks();
    });

    test('react function attr:data-qa', () => {
        act(() => {
            creator(WmlAttrsWasabyRoot, { dataQaFn: true }, container);
        });

        expect(container).toMatchSnapshot();
    });

    test('react component attr:data-qa', () => {
        act(() => {
            creator(WmlAttrsWasabyRoot, { dataQaCls: true }, container);
        });

        expect(container).toMatchSnapshot();
    });
});
