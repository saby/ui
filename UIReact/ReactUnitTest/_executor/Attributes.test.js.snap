// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markup attributes data-component ставиться на корректный div 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    <div
      data-component="ReactUnitTest/_executor/resources/DataComponent/DataComponent"
      id="hasDataComponent"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`Markup attributes Валидация строки style 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    <div
      id="divWithStyle"
    />
  </div>
</div>
`;

exports[`Markup attributes Валидация строки style ie 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    <div
      style="grid-template-columns: 1fr;"
    />
  </div>
</div>
`;

exports[`Markup attributes Валидация строки style normal 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    <div
      style="grid-template-columns: 1fr;"
    />
  </div>
</div>
`;

exports[`attr: в wml react component attr:data-qa 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    <div
      data-qa="data-qa-react-component"
    >
      react component
    </div>
  </div>
</div>
`;

exports[`attr: в wml react function attr:data-qa 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    <div
      data-qa="data-qa-react-function"
    >
      react function
    </div>
  </div>
</div>
`;
